src/sc/sidechainTxsCommitmentGuard.cpp

```diff
@@ -0,0 +1,280 @@
+    // as adding may fail for reasons different than having a full subtree.
+    if (counterRef.cert < cbs.CERT_LIMIT) {
+        ++counterRef.cert;
+        counterRef.bwt += bt_list_len;
+        LogPrint("sc", "%s():%d - scTxsCommitment guard: successfully added CERT for sidechain scId[%s].\n",
+            __func__, __LINE__, cert.GetScId().ToString());
+        return true;
+    } else {
+        LogPrint("sc", "%s():%d - scTxsCommitment guard failed: too many CERT for sidechain scId[%s].\n",
+            __func__, __LINE__, cert.GetScId().ToString());
+        return false;
+    }
+}
+
+bool SidechainTxsCommitmentGuard::add(const CTransaction& tx, bool autoRewind)
+{
+    if (!tx.IsScVersion())
+        return true;
+
+    LogPrint("sc", "%s():%d adding tx[%s] to ScTxsCommitmentGuard\n", __func__, __LINE__, tx.GetHash().ToString());
+
+    const uint256& tx_hash = tx.GetHash();
+
+    bool addOK = true;
+    int addedFt = 0, addedBwtr = 0, addedCsw = 0;
+
+    for (unsigned int fwtIdx = 0; fwtIdx < tx.GetVftCcOut().size(); ++fwtIdx)
+    {
+        const CTxForwardTransferOut& ccout = tx.GetVftCcOut().at(fwtIdx);
+
+        if (!add_fwt(ccout))
+        {
+            LogPrintf("%s():%d Error adding fwt: tx[%s], pos[%d]\n", __func__, __LINE__,
+                tx_hash.ToString(), fwtIdx);
+            addOK = false;
+            break;
+        } else {
+            addedFt++;
+        }
+    }
+
+    if (addOK) {
+        for (unsigned int bwtrIdx = 0; bwtrIdx < tx.GetVBwtRequestOut().size(); ++bwtrIdx)
+        {
+            const CBwtRequestOut& ccout = tx.GetVBwtRequestOut().at(bwtrIdx);
+
+            if (!add_bwtr(ccout))
+            {
+                LogPrintf("%s():%d Error adding bwtr: tx[%s], pos[%d]\n", __func__, __LINE__,
+                    tx_hash.ToString(), bwtrIdx);
+                addOK = false;
+                break;
+            } else {
+                addedBwtr++;
+            }
+        }
+    }
+
+    if (addOK) {
+        for (unsigned int cswIdx = 0; cswIdx < tx.GetVcswCcIn().size(); ++cswIdx)
+        {
+            const CTxCeasedSidechainWithdrawalInput& ccin = tx.GetVcswCcIn().at(cswIdx);
+
+            if (!add_csw(ccin))
+            {
+                LogPrintf("%s():%d Error adding csw: tx[%s], pos[%d]\n", __func__, __LINE__,
+                    tx_hash.ToString(), cswIdx);
+                addOK = false;
+                break;
+            } else {
+                addedCsw++;
+            }
+        }
+    }
+
+    // Restore CBS to a valid state if we were not able to add any of the FT / BWTR / CSW
+    if (!addOK && autoRewind) {
+        rewind(tx, addedFt, addedBwtr, addedCsw);
+        
+        // Remove from the maps all the sidechains without entities
+        std::vector<uint256> aliveToBeRemoved;
+        for (auto [scid, cbsa]: cbs.cbsaMap) {
+            if ((cbsa.bwt == 0) && (cbsa.bwtr == 0) && (cbsa.cert == 0) && (cbsa.ft == 0))
```

Comment:
(very minor importance) Just a matter a taste, would you consider this:

```
for (auto it = cbs.cbsaMap.cbegin(); it != cbs.cbsaMap.cend(); )
{
    if ((it->second.bwt == 0) && (it->second.bwtr == 0) && (it->second.cert == 0) && (it->second.ft == 0))
    {
        cbs.cbsaMap.erase(it++);
    }
    else
        ++it;
}
```

in order to avoid creating a vector for keeping track of the items to be removed?

Feel free to ignore if you prefer current style.

Reply:
I reimplemented this section in a very similar way to avoid the allocation of the vectors.